### PR TITLE
Fix misc memleaks on tuple encode failure

### LINKED ISSUE #7939: Fiber region memory leaks on invalid input to DML function
Reproduced with Tarantool 2.11.0-entrypoint-697-gbab289cd88a7. Tarantool 2.10 is affected as well.

Passing data that can't be encoded in MsgPack to a DML function may result in a fiber region memory leak:

```Lua
local fiber = require('fiber')

function print_mem_used()
    print('Mem used: ' .. fiber.info()[fiber.self().id()].memory.used)
end

box.cfg{log_level = 'warn'}
box.schema.create_space('test')
box.space.test:create_index('pk')

print_mem_used()
pcall(box.space.test.insert, box.space.test,
      {2, string.rep('x', 1024 * 1024), require('buffer').ibuf()})
print_mem_used()

os.exit(0)
```

Output:

```
Mem used: 0
Mem used: 2
```

This happens, because [`lbox_encode_tuple_on_gc`](https://github.com/tarantool/tarantool/blob/bab289cd88a7f76bd1334853063066c43591fabe/src/box/lua/misc.cc#L52) doesn't clean up the region on failure.

This is a minor bug, because the leaked memory will be freed on a transaction commit, but it blocks #5665.

### PR DESCRIPTION
**Pull request by nshy**
Fixes misc places where failure to encode tuple leads to memory leaks.

Closes #7939

### CODE REVIEW

File: src/lua/msgpack.c [hunk 1]
```diff
@@ -197,99 +198,113 @@ translate_map_key_field(struct luaL_field *field, uint32_t hash,
...
 			    field->type == MP_STR)
 				translate_map_key_field(field,
 							lua_hashstring(L, -1),
 							translation);
-			luamp_encode_with_translation_r(L, cfg, stream, field,
-							level + 1, translation);
+			if (luamp_encode_with_translation_r(
+				L, cfg, stream, field, level + 1,
+				translation, NULL) != 0)
+				goto error;
```
Comment (locker) [inline src/lua/msgpack.c:269]: Please fix checkpatch.

File: src/lua/msgpack.c [hunk 2]
```diff
@@ -510,7 +536,13 @@ lua_msgpack_encode(lua_State *L)
 	mpstream_init(&stream, buf, ibuf_reserve_cb, ibuf_alloc_cb,
 		      luamp_error, L);
```
Comment (locker) [inline src/lua/msgpack.c:537]: Without using an x alloc function or setting an error flag here, this patch doesn't seem to be complete - `luamp_error` may raise.

File: test/box-luatest/gh_7939_fix_meamleak_on_tuple_encode_test.lua [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+        ret, err = box.execute('SELECT ?;', {{trouble}})
+        t.assert_equals(ret, nil)
+        t.assert_equals(tostring(err), errmsg)
+        t.assert_equals(before, mem_used())
+    end)
+end
+
+g.after_all(function(cg)
+    cg.server:drop()
+end)
```
Comment (locker) [inline test/box-luatest/gh_7939_fix_meamleak_on_tuple_encode_test.lua:187]: Nit: Please move the `after_all` hander to the top, near `before_all` and `after_each`.

File: changelogs/unreleased/gh-7939-fix-misc-memory-leaks-on-tuple-encoding-failure.md [hunk 4]
```diff
@@ -0,0 +1,3 @@
+## bugfix/box
+
+* Fix memory leaks in miscellaneous places on tuple encoding failure (gh-7939).
```
Comment (locker) [inline changelogs/unreleased/gh-7939-fix-misc-memory-leaks-on-tuple-encoding-failure.md:3]: Fix**ed**

**Review by locker — CHANGES_REQUESTED** [review]

**Comment by locker** [comment]
**DO NOT MERGE** before https://github.com/tarantool/small/pull/62.

File: src/lua/msgpack.c (hunk 1 above)
Comment (nshy) [reply src/lua/msgpack.c:269]: Fixed.

File: src/lua/msgpack.c [hunk 5]
```diff
@@ -197,99 +198,113 @@ translate_map_key_field(struct luaL_field *field, uint32_t hash,
...
+				struct mh_strnu32_t *translation,
+				enum mp_type *type_out)
 {
 	int top = lua_gettop(L);
 	enum mp_type type;
 	const char *data;
 	size_t data_len;
 
 restart: /* used by MP_EXT of unidentified subtype */
+	type = field->type;
```
[the remark this answers is not in the document]
Comment (nshy) [reply src/lua/msgpack.c:216]: Fixed.

File: src/lua/msgpack.c [hunk 6]
```diff
@@ -510,7 +536,13 @@ lua_msgpack_encode(lua_State *L)
 	mpstream_init(&stream, buf, ibuf_reserve_cb, ibuf_alloc_cb,
 		      luamp_error, L);
 
-	luamp_encode(L, cfg, &stream, 1);
+	if (luamp_encode(L, cfg, &stream, 1) != 0) {
+		if (index == 1)
```
[the remark this answers is not in the document]
Comment (nshy) [reply src/lua/msgpack.c:540]: Fixed.

File: changelogs/unreleased/gh-7939-fix-misc-memory-leaks-on-tuple-encoding-failure.md (hunk 4 above)
Comment (nshy) [reply changelogs/unreleased/gh-7939-fix-misc-memory-leaks-on-tuple-encoding-failure.md:3]: Fixed.

File: test/box-luatest/gh_7939_fix_meamleak_on_tuple_encode_test.lua (hunk 3 above)
Comment (nshy) [reply test/box-luatest/gh_7939_fix_meamleak_on_tuple_encode_test.lua:187]: Done.

File: src/lua/msgpack.c (hunk 2 above)
Comment (nshy) [reply src/lua/msgpack.c:537]: This will be done in the scope of issue #8679

**Review by locker — APPROVED** [review]

**Comment by locker** [comment]
Waiting review from the doc team.

File: changelogs/unreleased/gh-7939-fix-misc-memory-leaks-on-tuple-encoding-failure.md [hunk 7]
```diff
@@ -0,0 +1,4 @@
+## bugfix/box
+
+* Fixed memory leaks in miscellaneous places on tuple encoding
+  failure (gh-7939).
```
Comment (p7nov) [inline changelogs/unreleased/gh-7939-fix-misc-memory-leaks-on-tuple-encoding-failure.md:4]:
```suggestion
* Fixed several memory leaks on tuple encoding failures (gh-7939).
```

**Review by p7nov — APPROVED** [review]
Please see my suggestion on the changelog.

File: changelogs/unreleased/gh-7939-fix-misc-memory-leaks-on-tuple-encoding-failure.md (hunk 7 above)
Comment (nshy) [reply changelogs/unreleased/gh-7939-fix-misc-memory-leaks-on-tuple-encoding-failure.md]: Done.

**Comment by locker** [comment]
ASAN [fails](https://github.com/tarantool/tarantool/actions/runs/5065905241/jobs/9095028778?pr=8680). Please fix.

**Comment by nshy** [comment]
There was an issue with ASAN workflow. Test `test/app-tap/msgpack.test.lua` failed because of leak detected.
The issue fixed by workaround:
```diff
diff --git a/test/app-tap/lua/serializer_test.lua b/test/app-tap/lua/serializer_test.lua
index 0a99d420f..b963c6462 100644
--- a/test/app-tap/lua/serializer_test.lua
+++ b/test/app-tap/lua/serializer_test.lua
@@ -441,12 +441,6 @@ local function test_depth(test, s)
     for _ = 1, max_depth + 1 do t = {t} end
     local ok = pcall(s.encode, t)
     test:ok(not ok, "too deep encode depth")
+    -- In our CI normally LSAN do not report as a leak the memory allocated
+    -- using malloc due to suppression 'leak:libc.so*' in asan/lsan.supp.
+    -- But unexpectedly if stack trace depth is large as in this case
+    -- (>128 frames) then suppression does not work and CI breaks. Let's
+    -- run garbage collector to avoid this particular leak.
+    collectgarbage()

     s.cfg({encode_max_depth = max_depth + 1})
     ok = pcall(s.encode, t)
```

Minimal proof that suppression failed to work with large stack trace depth:
```c
#include <stdlib.h>

static void
f(int d, int max)
{
        if (d < max)
                f(d + 1, max);
        else
                malloc(100);
}

int
main(int argc, char **argv)
{
        f(0, atoi(argv[1]));
        return 0;
}
```
```console
$ export | grep san
declare -x ASAN="ON"
declare -x ASAN_OPTIONS="heap_profile=0:unmap_shadow_on_exit=1:detect_invalid_pointer_pairs=1:symbolize=1:detect_leaks=1:dump_instruction_bytes=1:print_suppressions=0"
declare -x LSAN_OPTIONS="suppressions=<snap>/tarantool/asan/lsan.supp"
$ clang -fsanitize=leak -g tst.c
$ ./a.out 20
-----------------------------------------------------
Suppressions used:
  count      bytes template
      1        100 libc.so*
-----------------------------------------------------

$ ./a.out 100

=================================================================
==895013==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 100 byte(s) in 1 object(s) allocated from:
    #0 0x55aa74fb42e8 in __interceptor_malloc (/home/shiny/dev/tarantool/a.out+0x3d2e8) (BuildId: 00df92c4b70efc91190bdf41cb459fbdeeab9fdd)
    #1 0x55aa74fb5bc6 in f /home/shiny/dev/tarantool/tst.c:9:3
    #2 0x55aa74fb5bb7 in f /home/shiny/dev/tarantool/tst.c:7:3
    #3 0x55aa74fb5bb7 in f /home/shiny/dev/tarantool/tst.c:7:3
    #4 0x55aa74fb5bb7 in f /home/shiny/dev/tarantool/tst.c:7:3
    #5 0x55aa74fb5bb7 in f /home/shiny/dev/tarantool/tst.c:7:3
```

**Comment by locker** [comment]
Cherry-picked to 2.11.

**Pull request merged**